File: Green.Core/Interpreter.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Green
{
    public sealed class Interpreter
    {
        private readonly Reader _reader;
        private readonly Dictionary<string, object> _environment = new Dictionary<string, object>();
        private Evaluator _evaluator;

        public Interpreter()
        {
            _reader = new Reader();
            _environment.Add("+", new GreenFunction(Add));
            _evaluator = new Evaluator(_environment);
        }

        public object EvalSource(string source)
        {
            object result = null;
            foreach (var expr in _reader.Read(source))
                result = Eval(expr);
            return result;
        }

        private object Eval(ISyntax expr)
        {
            var compiler = new Compiler();
            var bytecode = compiler.Compile(expr);
            return _evaluator.Eval(bytecode);
        }

        public delegate object GreenFunction(object[] args);

        public static object Add(object[] args)
        {
            Int64 result = 0;
            foreach (var arg in args)
            {
                switch (arg)
                {
                    case Int64 intValue:
                        result += intValue;
                        break;
                    case UInt64 intValue:
                        result += (Int64)intValue;
                        break;
                    case Int32 intValue:
                        result += intValue;
                        break;
                    case UInt32 intValue:
                        result += intValue;
                        break;
                    default:
                        throw new RuntimeException($"+: bad argument {arg}");
                }
            }
            return result;
        }
    }
}

File: Green.Core/Interpreter.fs
namespace Green

open Read
open Compile
open Bytecode
open Obj
open Module

module Interpreter =

    type Interpreter() =

        let mainModule:GreenModule = GreenModule.empty "main"
        do GreenModule.tryImport mainModule "+" Prelude.``module`` "+" |> ignore
        do GreenModule.tryImport mainModule "=" Prelude.``module`` "=" |> ignore

        member __.Eval expr : Value =
            match compile expr with
            | None -> failwith "Some error in compiler"
            | Some block -> eval mainModule block

        member this.EvalSource (source:string) : Value =
            match read source with
            | UnexpectedEof -> raise (ReadError "Unexpected end of file")
            | UnexpectedRightBr -> raise (ReadError "Unexpected right bracket")
            | Success syntax -> List.fold (fun _ s -> this.Eval s) Value.empty syntax
